### ISSUE TITLE
Match/switch expression

### ISSUE THREAD

**Issue by 8LWXpg** [typst/typst#1316]
### Description

I would like to request a match/switch statement feature for Typst. This would allow me to write conditional expressions that match a value against a set of patterns and execute different code blocks accordingly. 

### Use Case

take Rust for example:

```
match x {
    x if x < 2 => println!("x is less than 2"),
    x if x < 4 => println!("x is less than 4"),
    x if x < 6 => println!("x is less than 6"),
    x if x < 8 => println!("x is less than 8"),
    _ => println!("x is greater than or equal to 8"),
}
```
it's better than writing a long if-else-if chain

**Comment by Enivex** [comment]
You can essentially achieve match statements in typst already

```
#{
  let x = 5
  ( (x < 2,"x is less than 2"),
    (x < 4,"x is less than 4"),
    (x < 6,"x is less than 6"),
    (x < 8,"x is less than 8"),
    (true, "no matches")).find(t => t.at(0)).at(1)
}
```

Outputs "x is less than 6"

**Comment by PgBiel** [comment]
A proper match statement could be made more powerful by using pattern matching; that could be really useful. Perhaps patterns could be extended to use some `#show`-like syntax, allowing something like `heading(level: 1, body: a @ "b") => ...` (this could even be reused for show rules themselves, but that might be going too far with the idea :p).

**Comment by AhmedSalih3d** [comment]
Is this not planned?

**Comment by Enivex** [comment]
> Is this not planned?

If it's not on the roadmap then there are no concrete plans, no. That doesn't mean it can't happen of course.

**Comment by Noobzik** [comment]
+1 on this feature, was writting a match expression just to found out it does not exists.
We have lambdas expressions but not patterns matching :/

**Comment by SoappGuy** [comment]
this probably should be in the roadmap 😁

**Comment by Mambouna** [comment]
Also just stumbled here after trying to find out how to write a match case... Would be great to have that.

**Comment by RigoLigoRLC** [comment]
"Typesetting in Rust" doesn't support the iconic Rusty pattern matching sounds overly hilarious
I personally had to complain and +1 on this feature

**Comment by odilf** [comment]
I'm currently taking a shot at implementing this! (assuming no one else is working on it)

In theory, a match expression wouldn't add anything novel, there is already destructuring (like in let bindings) so a match would go one-by-one in the arms and try to destructure it and evaluate the expression if it it can.

In terms of syntax, I see no reason to deviate from how rust does it. I think `if` guards could be very useful too, so I'm planning on also impementing that.

Hopefully a (draft) PR should be up in a few days!

**Comment by isuffix** [comment]
@odilf I don’t think anyone’s working on this, but you should probably look through the Discord forges for some previous discussions on match expressions.

I would definitely be careful about your expectations on this and keep them tempered. Match expressions are a pretty interconnected feature that’s dependent on the syntax+semantics of pattern matching, which is further dependent on the overall type system of Typst, which is going to be getting a large revamp with the unification of values and content coming soon. Getting match expressions implemented is less difficulty of implementation and more about how the semantics of those other parts of the language combine. You should really join the discord and drive some discussion about semantics if you want to make sure a PR would get merged.

But that shouldn’t stop you from seeing what an implementation would require or trying to learn the codebase! This can still be helpful for ironing out the semantics. If you need help or want a tour of anything in the codebase from lexing/parsing through to evaluation, I’d be happy to help out and answer questions!

**Comment by odilf** [comment]
Oh yeah now that you mention it I do remember seeing the type system revamp discussions... 

However, from what I can tell I think match expressions don't add anything novel? I have an initial implementation working that basically just calls `destructure`. The way that destructuring works might be changed in the revamp (I assume) but I think the relation of `destructure` to matching will stay the same. 

<details>
<summary> This is the code so far (not final of course, I'm experimenting and learning the codebase) </summary>

```rust
fn eval(self, vm: &mut Vm) -> SourceResult<Self::Output> {
    let matchend = self.matchend().eval(vm)?;
    for arm in self.arms() {
        // XXX: Is scopes necessary?
        vm.scopes.enter();

        if let Err(_) = destructure(vm, arm.pattern(), matchend.clone()) {
            continue;
        };

        if let Some(guard) = arm.guard() {
            let result = guard.eval(vm)?;
            match result {
                Value::Bool(true) => (),
                Value::Bool(false) => continue,
                _ => bail!(self.span(), "match guard didn't evaluate to a boolean"),
            }
        }

        let value = arm.expr().eval(vm)?;
        vm.scopes.exit();
        return Ok(value);
    }

    // TODO: Should this fail?
    Ok(Value::None)
}
```
</details>

I guess it's confusing because the feature is match expressions, but typst already has patterns and pattern matching! This is mostly plumbing it together from my end. Or at least that's my understanding, but of course I might just be missing something since I haven't thought too deeply about this!

Either way, I'll discuss it on the discord, and I will also keep myself humble and my expectations in check 😇. Thanks for the feedback!

**Comment by tingerrr** [comment]
I don't think the type rework would block any of this either, especially since it's not too dissimilar to something we already have from an implementation perspective: if-else-expressions. Since Typst is very dynamic, we're not held back by complexities like exhaustiveness analysis.

**Comment by laurmaedje** [comment]
I think it is quite connected to the type and styling systems, at least in the way I've imagined it. My explorations about this looked like this:

<img width="1172" height="1900" alt="Image" src="https://github.com/user-attachments/assets/4bf26313-2f9e-4822-9bcf-50eb6aa7110e" />

It's been a while since I posted this image on Discord and some stuff would likely be different, but I still like the design where types and selectors are basically the same thing, and match works like show rules.

**Comment by Andrew15-5** [comment]
One thing that I immediately notice is the absence of conditional pattern matching:

https://github.com/typst/typst/blob/78355421ad73fdcbe93b4acca890b439c4b6f98d/crates/typst-library/src/text/smartquote.rs?plain=1#L238-L269

**Comment by odilf** [comment]
Ok, so it seems there are various levels to a match implementation. I'll try to consolidate a bit what I'm thinking:

Firstly, there's just reserving the match keyword. I think we should do this regardless because it's pretty clear that we will want some kind of match expression eventually, and this way we prevent breakage and churn (I guess in theory we could do a warning period and then do the change to let library authors change conflicting variable names. I don't know if there's a way to quantify the impact of the change. My closest proxy is that in test suite there was exactly one test that broke because it used `match` as a variable name)

---

The next level up is a simple match that uses the existing desctructuring semantics, which is what I had implicitly assumed. This is a pretty limited version of pattern matching, but I think any potential design for "complete" pattern matching would include this version (i.e., as the wildcards [^1]), so in that sense this is a minimal subset of full pattern matching that can be relatively quickly shipped. I'll call it "mvp match" ("minimal viable product", following rust's conventions). I'm assuming this version has pattern alternatives (`or`s) and `if` guards, since those are relatively easy to implement.

[^1]: In other words, I think that in any design of match it's clear that you should be able to do `match x { 2 => "literal", (x, y) => "array of two", x => "anything else" }`. I don't think it's so clear on how it relates with custom type matching, maybe `match x { (num: x, string: y) => "array of number and string" }`?

**The main question is whether this mvp match is too limited to warrant implementing it in this state**, as far as I understand.


The argument of it being limited is that, at the end of the day, this version of match doesn't _really_ provide anything new, and can be mostly recreated with a chain of `if`/`else`s. Maybe you could even argue that calling it "pattern matching" is a misnomer since it's barely matching anything.

Personally, I think there is _some_ use for it. For example, it could implement the match statement of the comment above. It is also what is shown in the original comment at the top of the thread. Generally, even the mvp match is still an enhanced switch, and switches can be nice to have by themselves! It is arguably nicer syntax and more clear/concise than an `if` chain, especially when destructuring arrays. This mvp version is in fact what I wanted/expected when I first realized `match` was missing from typst and found this issue. 

```rust
// Original example at the top of the thread, as the proposed typst version
match x {
    x if x < 2 => "x is less than 2",
    x if x < 4 => "x is less than 4",
    x if x < 6 => "x is less than 6",
    x if x < 8 => "x is less than 8",
    _ => "x is greater than or equal to 8",
}
```

And the destructuring patterns are patterns, albeit relatively simple ones, but I think it warrants giving it the "match" name (and also for future expansion). 

---

And finally there's the more complete pattern matching, like what @laurmaedje showed above, which I don't think is a first issue candidate and not really what I was planning on implementing 😅. As @isuffix said, that needs more traversal considerations of what patterns are, where are allowed, what their relationship is with custom types and so on...

---

Personally, I like the "mvp" concept and I think it's worth to have this version of match expressions, but of course I understand that it might feel half-baked and that you prefer not to ship it until it's more complete. I might have missed some more downsides of shipping the limited match, since I'm admittedly biased. 

The decision is up to the community! (and, ultimately, @laurmaedje). I totally understand of course if we decide against the mvp match in the end. 

(also thank you everyone for reading this and giving feedback ^^)

**Comment by laurmaedje** [comment]
The original example is a bit of a rare use of match as essentially a chain of ifs. I think it serves only a small number of the matching use cases. I would not want to ship something that can only do this. 

More generally, while I'm overall favorable towards adding a pattern matching construct, it's also not necessarily a given and in particular not that it's more or less a direct copy of Rust's system. Something else might fit better with Typst.

TDLR: I believe a potential pattern matching construct still needs a _lot_ of design work and is not ready for implementation. I would really recommend you to look for something more approachable as a first issue candidate.

**Comment by Kile-Asmussen** [comment]
Just to not let this idea languish in discord, the conversation about a match expression I think boils down to two use cases:

- A better way to multiplex on types/values of inputs for the purposes of ad-hoc polymorphism
- A better way to write selectors/an expansion to selector utilities

With these two problems remedied, a match-statement is basically just a syntactic convenience.

### Multiplexing/non-duck typing polymorphism

For the purpose of multiplexing, I think the sorry state of `type(...)` and element-introspection though `.func()` should be remedied first. Might raise this as an actual issue.

- There are a few types that don't seem to be named in `std` but I might be misremembering.
- Using `.func()` there are several element functions also not named, and appear pretty much undocumented, such as `sequence` and `space`.
- Additionally, `std.symbol` is a type but `[#sym.AA].func()` is seemingly an element function that is named the same and takes the same arguments?
- Other cases exist, I'm sure.

Basically there should be a foolproof way to introspect _exactly_ what a given value is. There is currently pretty significant holes in this functionality which should be the bare minimum. Basically typst is at the level of _Lua_ (i.e. awful) and should be at least a little above that.

(Going further, one could argue there is a certain hierarchy of element functions. Especially apparent in the visualization elements. A `square` is inarguably a special case of `rect`, for instance. It is dubious whether implementing this hierarchy in a sensible manner is possible.)

Once that is done, the problem then becomes that writing:
```typst
#if type(x) == std.int and x < 10 {
   ...
}
```
just to avoid a type error is cumbersome, but whatever. It is uncommon enough to need to multiplex in this manner (basically only in library code) that this is second-priority.

(A simple `.is(<type>)` method as an alias, would suffice to remove a good bit of boilerplate. Overloading the `in` operator is also an option. `x in int` is arguably elegant and very set-theory-esque.)

(Alternatively if there was a language construct that could "catch" a type error in a boolean expression convert it to `false`, that could similarly solve some of this verbosity problem. This could in theory be a built in function that accepts a value and a test-function that either returns a boolean or panics, calling that function, and defaulting to false if the function panics.)

### Selector improvements

First, I think that pattern matching with variable capture is a luxury that could be afforded to go without. In pretty much every case where destructuring is useful, simply addressing the field or index possible, and the only savings is syntax.

What I don't think can be afforded to go without is that selectors are currently opaque objects. Their ability to match a piece of content is entirely constrained to show rules.

It would almost certainly have to require `context`, but a way of directly checking if a given piece of content matches a selector would be very useful for scripting, as it would basically enable a match expression over content to be implemented as a function kinda like:
```typst
#let matches_selector(x, sel, fn, ..rest) = {
  context if sel.matches(x) {
    fn(x)
  } else if rest.pos().len() >= 2 {
    matches_selector(x, ..rest)
  } else {
    none
  }
}
```
(Again, overloading `in` I think would be appropriate as a selector is kind of a subset?)

This would not be enough for general pattern matching on _all_ types similar to how pattern matching works in Ruby, but the astute reader will notice that with custom element functions/types essentially serving as user defined structured data types, this is essentially a match statement that works on user defined data types!

It would also enable more complex show rules that could depend on the content of whatever they are styling, by exploring the tree of child objects and comparing them to a selector, then making decisions based on that.

**Aside:** another _extremely convenient_ feature that is IMO needed in selectors is selecting over a range of values. This:
```typst
#show heading.where(level: 3).or(heading.where(level: 4), heading.where(level: 5)): set text(fill: red)
```
should not be as cumbersome to specify as it currently is. It should be more akin to:
```typst
#show heading.where(level: 3..=5): set text(fill: red)
```
or similar. More thought is needed for this, but the ability to select on sets of properties for brevity would be a major boon, and would enable the above `matches_selector` to work much better.

Whether this would require implementing range as a first-class type is up in the air, but that would probably be a good idea at least. At least some kind of genericized does-this-value-fit-here similar to Ruby's now partially outdated `===` operator.

### Conclusion

With (what appears to me to be) very little effort, the desirable properties of a proper match statement can be achieved, though with some verbosity:

- A more comprehensive suite of "taste testing" tools, in particular named types/element functions in the standard library, bringing typst up to at least the level of e.g. Python.
- A "does this content match my selector" method which would then be usable with custom elements byt matching on their constructors.

(The latter is _technically_ possible to omit with the tradeoff of incurring lots more of the former, but at that point, we're not talking about pattern matching.)

Using these tools, a `match` statement begins to truly look like syntactic sugar, and an early version of the match statement could behind the scenes be implemented as a simple syntax rewriting rule. (Or not, idk how the typst compiler works in particular.)

The rest is bikeshedding.
